src/theory/quantifiers/sygus_grammar_norm.h

```diff
@@ -14,137 +14,439 @@
+ *   A -> B + B | x; B -> 0 | 1
+ * - consider the case in which in which the operator occurs nested in an
+ *   argument type of itself, e.g.:
+ *   A -> (B + B) + B | x; B -> 0 | 1
+ */
+class TransfChain : public Transf
+{
+ public:
+  TransfChain(unsigned chain_op_pos, std::vector<unsigned> elem_pos)
+      : d_chain_op_pos(chain_op_pos), d_elem_pos(elem_pos){};
+
+  /** builds types encoding a chain in which each link contains a repetition of
+   * the application of the chain operator over a non-identity element
+   *
+   * Example: when considering, over the integers, the operator "+" and the
+   * elemenst "1", "x" and "y", the built chain is e.g.
+   *
+   * x + ... + x + y + ... + y + 1 + ...+ 1
+   *
+   * whose encoding in types would be e.g.
+   *
+   * A  -> AX | AX + A | B
+   * AX -> x
+   * B  -> BY | BY + B | C
+   * BY -> y
+   * C  -> C1 | C1 + C
+   * C1 -> 1
+   *
+   * ++++++++
+   *
+   * The types composing links in the chain are built recursively by invoking
+   * sygus_norm, which caches results and handles the global normalization, on
+   * the operators not used in a given link, which will lead to recalling this
+   * transformation and so on until all operators originally given are
+   * considered.
+   */
+  virtual void buildType(SygusGrammarNorm* sygus_norm,
+                         TypeObject& to,
+                         const Datatype& dt,
+                         std::vector<unsigned>& op_pos) override;
+
+  /** Which sort of operators to look for in a type to determine if amenable for
+   * this transformation
+   *
+   * OP indicates which operator (e.g. PLUS for Int)
+   * ELEMS_ID indicates which element is the identitiy for the chain operator
+   * (e.g. ZERO for Int)
+   */
+  enum Block
+  {
+    OP,
+    ELEMS_ID
+  };
+
+  /** Specifies for each type node which are the elements of each block
+   *
+   * For example, for Int the map is {OP -> +, ELEMS_ID -> 0}
+   */
+  static std::map<TypeNode, std::map<Block, Node>> d_assoc;
+
+  /* Retrives, or, if none, creates, stores and returns, the map from block to
+   * nodes for the given type */
+  static inline std::map<Block, Node> getTypeAssoc(TypeNode tn)
+  {
+    auto it = d_assoc.find(tn);
+    if (it == d_assoc.end())
+    {
+      return addTypeAssoc(tn);
+    }
+    return it->second;
+  }
+
+ private:
+  /* Position of chain operator */
+  unsigned d_chain_op_pos;
+  /* Positions (of constructors in the datatype whose type is being normalized)
+   * of elements the chain operator is applied to */
+  std::vector<unsigned> d_elem_pos;
 
-/** Utility for normalizing SyGuS grammars and avoid spurious enumarations
+  /** associates elements to the transformation blocks for the given type
+   * node. Which elements these are is defined here.
+   *
+   * returns the built map
+   *
+   * TODO: #1304: Cover more types, make this robust to more complex grammars
+   */
+  static inline std::map<Block, Node> addTypeAssoc(TypeNode tn)
```

Comment:
Types can have more than one chainable operator. It looks as though this is assuming there can only be one chainable operator per type.